# 01. Importing libraries

In [1]:
# Import libraries
import warnings
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [2]:
# Define csv path
path = r'C:\Users\lucav\Desktop\Instacart Basket Analysis'

In [3]:
# Import data from orders.csv
warnings.simplefilter(action='ignore', category=FutureWarning)
df_ords = pd.read_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'), index_col = 0)

In [4]:
# Import data from products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 03. Data consistency checks

## Missing values

In [5]:
# Find missing values in df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [6]:
# Create a subset of df_prods containing only missing values
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [7]:
# Return df_nan
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [8]:
# Return the number of rows and columns in df_prods
df_prods.shape

(49693, 5)

In [9]:
# Create a subset of df_prods containing only non-missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]

In [10]:
# Return the number of rows and columns in df_prods_clean
df_prods_clean.shape

(49677, 5)

## Duplicate values

In [11]:
# Find duplicate values in df_prods_clean
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [12]:
# Return df_dups
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [13]:
# Return the number of rows and columns in df_prods_clean
df_prods_clean.shape

(49677, 5)

In [14]:
# Create a subset of df_prods_clean containing only non-duplicate values
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [15]:
# Return df_prods_clean_no_dups
df_prods_clean_no_dups.shape

(49672, 5)

## Export 'df_prods_clean_no_dups' as 'products_checked.csv'

In [16]:
df_prods_clean_no_dups.to_csv(os.path.join(path,'02 Data','Prepared Data','products_checked.csv'))

# 04. Task

## I. Perform consistency checks on df_prods

Completed. See above '03. Data consistency checks'

## II. Look for anomalies in df_prods

In [17]:
# Run the df.describe() function on df_prods
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The max value of 'prices' column is 99999. This value is probably incorrect and would require further investigations.

## III. Mixed-type data

In [18]:
# Check for mixed-type data in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## IV. If you find mixed-type data, fix it

No mix-type data found.

## V. Missing values

In [19]:
# Find missing values in df_ords
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
day_of_week                    0
hour_of_day                    0
days_since_prior_order    206209
dtype: int64

In [20]:
# Create a subset of df_ords containing only missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [21]:
# Return df_ords_nan
df_ords_nan

order_id  user_id  order_number  day_of_week  hour_of_day  \
0         2539329        1             1            2            8   
11        2168274        2             1            2           11   
26        1374495        3             1            1           14   
39        3343014        4             1            6           11   
45        2717275        5             1            3           12   
...           ...      ...           ...          ...          ...   
3420930    969311   206205             1            4           12   
3420934   3189322   206206             1            3           18   
3421002   2166133   206207             1            6           19   
3421019   2227043   206208             1            1           15   
3421069   3154581   206209             1            3           11   

         days_since_prior_order  
0                           NaN  
11                          NaN  
26                          NaN  
39                          NaN  
45                          NaN  
...                         ...  
3420930                     NaN  
3420934                     NaN  
3421002                     NaN  
3421019                     NaN  
3421069                     NaN  

[206209 rows x 6 columns]

In [22]:
# Run the df.describe() function on df_ords_nan
df_ords_nan.describe()

order_id        user_id  order_number    day_of_week  \
count  2.062090e+05  206209.000000      206209.0  206209.000000   
mean   1.708462e+06  103105.000000           1.0       2.754118   
std    9.881299e+05   59527.555167           0.0       2.076205   
min    2.000000e+01       1.000000           1.0       0.000000   
25%    8.507300e+05   51553.000000           1.0       1.000000   
50%    1.706246e+06  103105.000000           1.0       3.000000   
75%    2.564292e+06  154657.000000           1.0       5.000000   
max    3.421081e+06  206209.000000           1.0       6.000000   

         hour_of_day  days_since_prior_order  
count  206209.000000                     0.0  
mean       13.626597                     NaN  
std         4.223769                     NaN  
min         0.000000                     NaN  
25%        11.000000                     NaN  
50%        14.000000                     NaN  
75%        17.000000                     NaN  
max        23.000000                     NaN

206209 observations display 'NaN' in the 'days_since_prior_order' column. Looking at the result of the describe() function, it can be noted that all the 'NaN' observations correspond to a 'order_number' of 1; this indicates that a **'NaN' value is assigned to the first order placed by each customer.**

## VI. Address the missing values using an appropriate method

In [23]:
# Create a new column to display whether an order is the first
df_ords['first_order'] = df_ords['days_since_prior_order'].isnull() == True

In [24]:
# Return the first 5 rows of df_ords
df_ords.head()

order_id  user_id  order_number  day_of_week  hour_of_day  \
0   2539329        1             1            2            8   
1   2398795        1             2            3            7   
2    473747        1             3            3           12   
3   2254736        1             4            4            7   
4    431534        1             5            4           15   

   days_since_prior_order  first_order  
0                     NaN         True  
1                    15.0        False  
2                    21.0        False  
3                    29.0        False  
4                    28.0        False

An 'NaN' value corresponds to the first order ever placed by each customer. Thus, a new variable has been created to act like a flag based on the missing value, returning 'True' when an order is the first and 'False' when it isn't.

## VII. Duplicate values

In [25]:
# Find duplicate values in df_ords
df_ords_dups = df_ords[df_ords.duplicated()]

In [26]:
# Return df_ords_dups
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, day_of_week, hour_of_day, days_since_prior_order, first_order]
Index: []

## VIII. Address the duplicates using an appropriate method

No duplicate values found.

## IX. Export 'df_ords' as 'orders_checked.csv'

In [27]:
df_ords.to_csv(os.path.join(path,'02 Data','Prepared Data','orders_checked.csv'))